### PR TITLE
Use builtins adapted from cilium

### PR DESCRIPTION
**Pull request by usamasaqib**
### What does this PR do?
This PR adds builtins which take advantage of the fact that map values and context pointers, etc, are guaranteed to be aligned, to generate efficient bytecode.

### Motivation
Assuming r0 points to some memory location, and r2 points to a stack variable, when using `__builtin_memcpy` to copy data the compiler generates the following bytecode:
```
......
// Copy r0[30] to r2[0]
r1 = *(u8 *)(r0 + 30)
*(u8 *)(r2 + 30) = r1
......
// Copy r0[0] to r2[0]
r1 = *(u8 *)(r0 + 0)
*(u8 *)(r2 + 0) = r1
```
This bytecode copies the memory from r0 to r2, byte by byte.

For an object for 247 bytes, copying to stack requires 494 instructions. 1 million executions of a socket filter would result in 247 million memory read and 247 million memory write operations when using `__builtin_memcpy`.

The optimized `bpf_memcpy` introduced in this PR, [taken from the cilium project](https://github.com/cilium/cilium/blob/master/bpf/include/bpf/builtins.h#L152), makes assumptions about memory alignment enforced by the [verifier](https://elixir.bootlin.com/linux/v5.15/source/kernel/bpf/verifier.c#L3567), to generate optimized bytecode for copying. For any read of `x` bytes, the worst case scenario is when `x % 8 == 7`. In this case the number of instructions for copying the memory is given by the formula:
```
 2 ( 3 + ((x-7)/8) )
 |   |      |_ (x-7)/8 8-byte copies remain.
 |   |_ Three copies of size 8, 16 and 32, to align on qword boundary
 |_ Each copy of upto 64 bytes will generate two instructions.
```
The optimized `bpf_memcpy` will do enough copies to align on 8 byte boundary, and then copy 8 bytes at a time.
For the same object of size 247 bytes. Copying using the optimized helper will take 33 memory reads and 33 memory writes for a total of 66 instructions. Over 1 million invocations this reduces the instructions executed to 66 million, a reduction of more than 87%.
 

### Additional Notes
For the cases where we are copying a stack value to a memory location, the compiler is smarter and generates efficient bytecode even when using `__builtin_memcpy`

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the `k8s/<min-version>` label, indicating the lowest Kubernetes version compatible with this feature. 
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Comment by brycekahle** [comment]
Two things:
1. If we are still working on the results/justification for this change, let's move it back to Draft status.
2. If we pulled code from cilium, let's ensure we retain their copyright/license comments.

**Review by hmahmood — APPROVED** [review]

File: pkg/network/ebpf/c/runtime/conn-tuple.h [hunk 1]
```diff
@@ -106,7 +106,7 @@ static __always_inline int read_conn_tuple_partial(conn_tuple_t* t, struct sock*
  * Reads values into a `conn_tuple_t` from a `sock`. Initializes all values in conn_tuple_t to `0`. Returns 1 success, 0 otherwise.
  */
 static __always_inline int read_conn_tuple(conn_tuple_t* t, struct sock* skp, u64 pid_tgid, metadata_mask_t type) {
-    __builtin_memset(t, 0, sizeof(conn_tuple_t));
+    bpf_memset(t, 0, sizeof(conn_tuple_t));
```
Comment (brycekahle) [inline pkg/network/ebpf/c/runtime/conn-tuple.h:109]: This file should include `bpf_builtins.h`

File: pkg/network/ebpf/c/runtime/conntrack.c [hunk 2]
```diff
@@ -3,6 +3,7 @@
 
 #include "bpf_telemetry.h"
 #include "bpf_endian.h"
+#include "bpf_builtins.h"
```
Comment (brycekahle) [inline pkg/network/ebpf/c/runtime/conntrack.c:6]: Excess include?

File: pkg/network/ebpf/c/prebuilt/sock.h [hunk 3]
```diff
@@ -258,7 +258,7 @@ static __always_inline int read_conn_tuple_partial(conn_tuple_t * t, struct sock
  * Reads values into a `conn_tuple_t` from a `sock`. Initializes all values in conn_tuple_t to `0`. Returns 1 success, 0 otherwise.
  */
 static __always_inline int read_conn_tuple(conn_tuple_t* t, struct sock* skp, u64 pid_tgid, metadata_mask_t type) {
-    __builtin_memset(t, 0, sizeof(conn_tuple_t));
+    bpf_memset(t, 0, sizeof(conn_tuple_t));
```
Comment (brycekahle) [inline pkg/network/ebpf/c/prebuilt/sock.h:261]: This file should include `bpf_builtins.h`

File: pkg/network/ebpf/c/prebuilt/offset-guess.c [hunk 4]
```diff
@@ -1,6 +1,7 @@
 #include "kconfig.h"
 #include "offset-guess.h"
 #include "bpf_helpers.h"
+#include "bpf_builtins.h"
```
Comment (brycekahle) [inline pkg/network/ebpf/c/prebuilt/offset-guess.c:4]: Excess include?

File: pkg/network/ebpf/c/ip.h [hunk 5]
```diff
@@ -53,7 +53,7 @@ static __always_inline void read_ipv4_skb(struct __sk_buff *skb, __u64 off, __u6
 // On older kernels, clang can generate Wunused-function warnings on static inline functions defined in 
 // header files, even if they are later used in source files. __maybe_unused prevents that issue
 __maybe_unused static __always_inline __u64 read_conn_tuple_skb(struct __sk_buff *skb, skb_info_t *info, conn_tuple_t *tup) {
-    __builtin_memset(info, 0, sizeof(skb_info_t));
+    bpf_memset(info, 0, sizeof(skb_info_t));
```
Comment (brycekahle) [inline pkg/network/ebpf/c/ip.h:56]: This file should include `bpf_builtins.h`

File: pkg/network/ebpf/c/https.h [hunk 6]
```diff
@@ -17,8 +17,8 @@ static __always_inline int http_process(http_transaction_t *http_stack, skb_info
 
 static __always_inline void https_process(conn_tuple_t *t, void *buffer, size_t len, __u64 tags) {
     http_transaction_t http;
-    __builtin_memset(&http, 0, sizeof(http));
-    __builtin_memcpy(&http.tup, t, sizeof(conn_tuple_t));
+    bpf_memset(&http, 0, sizeof(http));
```
Comment (brycekahle) [inline pkg/network/ebpf/c/https.h:20]: This file should include `bpf_builtins.h`

File: pkg/network/ebpf/c/http.h [hunk 7]
```diff
@@ -71,7 +71,7 @@ static __always_inline void http_enqueue(http_transaction_t *http) {
         return;
     }
 
-    __builtin_memcpy(&batch->txs[batch->pos], http, sizeof(http_transaction_t));
+    bpf_memcpy(&batch->txs[batch->pos], http, sizeof(http_transaction_t));
```
Comment (brycekahle) [inline pkg/network/ebpf/c/http.h:74]: This file should include `bpf_builtins.h`

File: pkg/network/ebpf/c/conntrack.h [hunk 8]
```diff
@@ -40,7 +40,7 @@ static __always_inline void print_translation(const conntrack_tuple_t *t) {
 }
 
 static __always_inline int nf_conntrack_tuple_to_conntrack_tuple(conntrack_tuple_t *t, const struct nf_conntrack_tuple *ct) {
-    __builtin_memset(t, 0, sizeof(conntrack_tuple_t));
+    bpf_memset(t, 0, sizeof(conntrack_tuple_t));
```
Comment (brycekahle) [inline pkg/network/ebpf/c/conntrack.h:43]: This file should include `bpf_builtins.h`

File: pkg/network/ebpf/c/http-buffer.h [hunk 9]
```diff
@@ -10,7 +10,7 @@
 //
 // This function is used for the uprobe-based HTTPS monitoring (eg. OpenSSL, GnuTLS etc)
 static __always_inline void read_into_buffer(char *buffer, char *data, size_t data_size) {
-    __builtin_memset(buffer, 0, HTTP_BUFFER_SIZE);
+    bpf_memset(buffer, 0, HTTP_BUFFER_SIZE);
```
Comment (brycekahle) [inline pkg/network/ebpf/c/http-buffer.h:13]: This file should include `bpf_builtins.h`

File: pkg/network/ebpf/c/runtime/skb.h [hunk 10]
```diff
@@ -29,7 +29,7 @@ static __always_inline int sk_buff_to_tuple(struct sk_buff *skb, conn_tuple_t *t
     }
 
     struct iphdr iph;
-    __builtin_memset(&iph, 0, sizeof(struct iphdr));
+    bpf_memset(&iph, 0, sizeof(struct iphdr));
```
Comment (brycekahle) [inline pkg/network/ebpf/c/runtime/skb.h:32]: This file should include `bpf_builtins.h`

File: pkg/network/ebpf/c/sockfd.h [hunk 11]
```diff
@@ -28,7 +28,7 @@ __maybe_unused static __always_inline void clear_sockfd_maps(struct sock* sock)
 
     // Copy map value to stack before re-using it (needed for Kernel 4.4)
     pid_fd_t pid_fd_copy = {};
-    __builtin_memcpy(&pid_fd_copy, pid_fd, sizeof(pid_fd_t));
+    bpf_memcpy(&pid_fd_copy, pid_fd, sizeof(pid_fd_t));
```
Comment (brycekahle) [inline pkg/network/ebpf/c/sockfd.h:31]: This file should include `bpf_builtins.h`

File: pkg/network/ebpf/c/tracer-events.h [hunk 12]
```diff
@@ -92,7 +92,7 @@ static __always_inline void flush_conn_close_if_full(struct pt_regs *ctx) {
         // This is necessary for older Kernel versions only (we validated this behavior on 4.4.0),
         // since you can't directly write a map entry to the perf buffer.
         batch_t batch_copy = {};
-        __builtin_memcpy(&batch_copy, batch_ptr, sizeof(batch_copy));
+        bpf_memcpy(&batch_copy, batch_ptr, sizeof(batch_copy));
```
Comment (brycekahle) [inline pkg/network/ebpf/c/tracer-events.h:95]: This file should include `bpf_builtins.h`

File: pkg/network/ebpf/c/tracer-stats.h [hunk 13]
```diff
@@ -15,7 +15,7 @@ static __always_inline u32 get_sk_cookie(struct sock *sk) {
 static __always_inline conn_stats_ts_t *get_conn_stats(conn_tuple_t *t, struct sock *sk) {
     // initialize-if-no-exist the connection stat, and load it
     conn_stats_ts_t empty = {};
-    __builtin_memset(&empty, 0, sizeof(conn_stats_ts_t));
+    bpf_memset(&empty, 0, sizeof(conn_stats_ts_t));
```
Comment (brycekahle) [inline pkg/network/ebpf/c/tracer-stats.h:18]: This file should include `bpf_builtins.h`

**Review by brycekahle — CHANGES_REQUESTED** [review]
We should try to follow "include what you use". It makes it simpler to reason about the requirements of specific files, and you don't get surprise compiler errors.

**Review by guyarb — CHANGES_REQUESTED** [review]
Agreed with @brycekahle comments

**Review by brycekahle — APPROVED** [review]

**Review by guyarb — APPROVED** [review]
LGTM

**Pull request merged**